### PR TITLE
Concretizer should respect namespace of reused specs

### PR DESCRIPTION
**Pull request by becker33**
TLDR: This PR fixes a bug in the concretizer by imposing `namespace` attributes from imposed specs.

In discussion on #45416 @brettviren reported a bug:

> I tested with a user repo (wirecell in examples below, https://github.com/wirecell/wire-cell-spack) and config.yaml like in your first comment on this issue. My repo provides a jsonnet (which is also in builtin) and a go-jsonnet package (which is unique).
>
> ```$ spack list -r wirecell jsonnet
> go-jsonnet  jsonnet
> ==> 2 packages
> $ spack list -r builtin jsonnet
> jsonnet
> ==> 1 packages
> $ spack list jsonnet
> go-jsonnet  jsonnet
> ==> 2 packages 
> ```
> The only major problem I see is when trying to concretize the jsonnet from builtin
>
> ```$ spack spec --format '{namespace} {name} {version}' jsonnet
> wirecell jsonnet 0.19.1
> $ spack spec --format '{namespace} {name} {version}' wirecell.jsonnet
> wirecell jsonnet 0.19.1
> $ spack spec --format '{namespace} {name} {version}' builtin.jsonnet
> ==> Error: Internal Spack error: the solver completed but produced specs that do not satisfy the request. Please report a bug at https://github.com/spack/spack/issues
> 	Unsatisfied input specs:
> 	Input spec: jsonnet
> 	Candidate spec: jsonnet@=0.19.1%gcc@=12.2.0~python build_system=makefile arch=linux-debian12-haswell > ^[deptypes=link] gcc-runtime@=12.2.0%gcc@=12.2.0 build_system=generic arch=linux-debian12-haswell ^[deptypes=build,link virtuals=iconv,libc] glibc@=2.36%gcc@=12.2.0 build_system=autotools arch=linux-debian12-haswell ^[deptypes=build] gmake@=4.4.1%gcc@=12.2.0~guile build_system=generic arch=linux-debian12-haswell
> ```

Through further debugging we determined (see https://github.com/spack/spack/pull/45416#issuecomment-2261197202) that reused specs can be assigned a namespace different than the one they were concretized with. This was possible because the `namespace` attribute was modeled in the solver, and was instead statically assigned to the highest priority repo.

Downsides to this approach: you can reuse a package that is not from the repo you expect.
Upsides to this approach: you can reuse packages even if you override irrelevant aspects of the package (like the tests) in a local repo.

### CODE REVIEW

**Comment by becker33** [comment]
@spackbot run pipeline

**Review by tgamblin — APPROVED** [review]
